### PR TITLE
Add support best value with constraint

### PR DESCRIPTION
**Pull request by hrntsm**
## Motivation

implement #5184 

## Description of the changes

Until now, when finding the value of best in Study, it was calculated from the COMPLETE trial. In addition to that, it was changed so that the trial would satisfy the constraint.

Also added a warn output to show the changes made.

### CODE REVIEW

File: optuna/storages/_base.py
Comment (hrntsm) [inline optuna/storages/_base.py:21]: I would like to use `_CONSTRAINT_KEY` in `optuna.sampler._base`, but it would be a circular import, so it is a direct value. Please let me know if there is a better way.

File: optuna/study/_multi_objective.py
Comment (hrntsm) [inline optuna/study/_multi_objective.py:12]: ditto

**Comment by nabenabe0928** [comment]
Could @not522 and @eukaryo review this PR?

**Comment by not522** [comment]
@nabenabe0928 I've been busy, so could you please reassign the reviewer?

**Comment by nabenabe0928** [comment]
Could you review this PR? @HideakiImamura

**Comment by HideakiImamura** [comment]
I would like to discuss the design of the implementation. As @not522 suggested [here](https://github.com/optuna/optuna/issues/5184#issuecomment-1891212936), I think we should not change the internal implementation of `BaseStorage` on the current situation because it changes the existing behavior for all users and downgrade the speed performance. How about implementing the experimental method `get_best_feasible_trial` or attribute `best_feasible_trial` of the `Study` and emitting the `Userwarning` to explain that the `best_feasible_trial` is slower than `best_trial`?

**Comment by hrntsm** [comment]
Thanks for confirming.
I will modify it with that information.

**Comment by hrntsm** [comment]
@HideakiImamura 
I implemented this by eliminating the changes to strorage._base and adding attributes to Study.
How about this one?
CI is failing in the mypy test section, but I will add a test if this direction is acceptable.

**Comment by HideakiImamura** [comment]
@hrntsm Sorry for the late reply. The current approach looks great to me. Please proceed this direction.

**Comment by hrntsm** [comment]
@HideakiImamura 
I fixed fails in CI. Could you please review?

File: optuna/study/study.py
Comment (HideakiImamura) [inline optuna/study/study.py:199]: Is this warning necessary? It's already pointed out in docstring, so I don't think there's a need to display it everytime it runs. From a different perspective, following [the coding style convention](https://github.com/optuna/optuna/wiki/Coding-Style-Conventions#logging-howto), if you just want to notify the user with a warning, I think you should use `Logger.warning()`.

File: optuna/study/study.py
Comment (HideakiImamura) [inline optuna/study/study.py:202]: Please add the test case for this error.

File: optuna/study/study.py
Comment (HideakiImamura) [inline optuna/study/study.py:208]: Ditto.

File: optuna/study/study.py
Comment (HideakiImamura) [inline optuna/study/study.py:217]: Ditto.

File: optuna/study/study.py
Comment (HideakiImamura) [inline optuna/study/study.py:219]: This check is already covered in L201. Can be omitted.

File: optuna/study/study.py
Comment (HideakiImamura) [inline optuna/study/study.py:214]: How about using `_get_feasible_trials` in `study/_multi_objective.py`?

File: optuna/study/study.py
Comment (HideakiImamura) [reply optuna/study/study.py:208]: This check is covered in L216, so can be omitted.

File: optuna/study/study.py
Comment (HideakiImamura) [inline optuna/study/study.py:193]: I think we should add the `Error` condition in the docstring.

File: optuna/study/study.py
Comment (HideakiImamura) [inline optuna/study/study.py:226]: Could you add the test case for this?

File: optuna/study/study.py
Comment (HideakiImamura) [inline optuna/study/study.py:246]: This warning can be removed.

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR. I have several comments. PTAL.

**Comment by HideakiImamura** [comment]
@hrntsm I've discussed with other commiters about the direction of this PR, but haven't reached a conclusion as the opinions diverged. However, since we've received opinions from many that we should be cautious about introducing new public APIs, there's a high chance that the current policy might change. Could you please hold off on making amendments to the PR for now?

**Comment by HideakiImamura** [comment]
We've had a discussion amongst the commiters regarding the direction of this PR. Given the significant impact of introducing a public API, we've concluded that it may be best to conduct it not in the upcoming minor version v3.6, but in the next planned major version v4.0. Would it be acceptable for this PR to only include changes in the private API, i.e., adding an argument to consider constraints only to `_get_pareto_front_trials`? Although we won't officially offer the API to users, users who really want to use this feature at this point will use the private API, knowing that the feature might change in the future. As for `best_feasible_trial`/`best_feasible_trials`, please allow us to reconsider it in v4.

**Comment by HiroakiNatsume** [comment]
@HideakiImamura 
Thank you for the confirmation within the committer.
I agree with this conclusion as I too had concerns about the impact.
As for the specific modifications, would it be correct to eliminate the changes in Study.py and only modify _multi_objective.py with the comments I received?

**Comment by hrntsm** [comment]
@HideakiImamura
Thank you for the confirmation within the committer.
I agree with this conclusion as I too had concerns about the impact.

As for the specific modifications, would it be correct to eliminate the changes in Study.py and only modify _multi_objective.py with the comments I received?

**Review by HideakiImamura — APPROVED** [review]
Thanks for the update! LGTM.

**Comment by eukaryo** [comment]
LGTM.

**Comment by hrntsm** [comment]
@eukaryo 

Thanks for the review. I have fixed the conflicts, please merge.

**Pull request merged**
